Index: web/WEB-INF/tags/tags/_tagsLinkPopup.tag
===================================================================
--- web/WEB-INF/tags/tags/_tagsLinkPopup.tag	(revision Shelved version)
+++ web/WEB-INF/tags/tags/_tagsLinkPopup.tag	(revision Shelved version)
@@ -0,0 +1,23 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"
+  %><%@ taglib prefix="bs" tagdir="/WEB-INF/tags"
+  %><%@ taglib prefix="tags" tagdir="/WEB-INF/tags/tags"
+  %><%@ attribute name="buildPromotion" fragment="false" required="true" type="jetbrains.buildServer.serverSide.BuildPromotion"
+  %>
+<c:set var="id" value="${buildPromotion.id}"/>
+<script type="text/javascript">
+  BS['_ViewTagsPopup${id}'] = new BS.Popup('viewTagsPopup-${id}', {
+    shift: {x: -20, y: 20}}
+  );
+</script>
+<c:set var="attrs" value="id='tags:${id}'"/>
+<bs:popupControl showPopupCommand="BS['_ViewTagsPopup${id}'].showPopupNearElement(this, ${id});"
+                 hidePopupCommand="BS['_ViewTagsPopup${id}'].hidePopup();"
+                 stopHidingPopupCommand="BS['_ViewTagsPopup${id}'].stopHidingPopup();"
+                 controlId="tags:${id}"><jsp:doBody/></bs:popupControl>
+<div id="viewTagsPopup-${id}" class="popupDiv">
+  <c:set var="editTagsButton">&nbsp;<tags:editTagsLink className="btn btn_mini" buildPromotion="${buildPromotion}">Edit</tags:editTagsLink></c:set>
+  <tags:showBuildPromotionTags buildPromotion="${buildPromotion}" hidePrivateTags="${true}">
+    <jsp:attribute name="onNoTags">There are no tags${editTagsButton}</jsp:attribute>
+    <jsp:attribute name="afterTags">${editTagsButton}</jsp:attribute>
+  </tags:showBuildPromotionTags>
+</div>
\ No newline at end of file
Index: web/WEB-INF/tags/buildRow.tag
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- web/WEB-INF/tags/buildRow.tag	(revision Local version)
+++ web/WEB-INF/tags/buildRow.tag	(revision Shelved version)
@@ -196,8 +196,8 @@
   </c:if
 
   ><c:if test="${showTags}">
-    <td class="${rowClass} tags">
-      <t:tagsInfo buildPromotion="${build.buildPromotion}"/>
+    <td class="${rowClass} tags" style="text-align: right; padding-right: 1.5em;">
+      <t:_tagsInfo buildPromotion="${build.buildPromotion}"/>
     </td>
   </c:if
 
Index: web/resources/css/tags.css
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- web/resources/css/tags.css	(revision Local version)
+++ web/resources/css/tags.css	(revision Shelved version)
@@ -47,3 +47,21 @@
   text-decoration: none !important;
 }
 
+.icon-tag,
+.icon-tags {
+  text-decoration: none !important;
+}
+
+.oneTag, .manyTags{
+  color: #739CB3;
+  cursor: pointer;
+  font-style: normal;
+}
+
+.noTags{
+  color: #739CB3;
+  opacity: 0.5;
+  cursor: pointer;
+  font-style: normal;
+}
+
Index: web/WEB-INF/tags/tags/_tagsInfo.tag
===================================================================
--- web/WEB-INF/tags/tags/_tagsInfo.tag	(revision Shelved version)
+++ web/WEB-INF/tags/tags/_tagsInfo.tag	(revision Shelved version)
@@ -0,0 +1,37 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"
+  %><%@ taglib prefix="fn" uri="http://java.sun.com/jsp/jstl/functions"
+  %><%@ taglib prefix="bs" tagdir="/WEB-INF/tags"
+  %><%@ taglib prefix="tags" tagdir="/WEB-INF/tags/tags"
+  %><%@ attribute name="buildPromotion" required="true" type="jetbrains.buildServer.serverSide.BuildPromotion"
+  %><%@ attribute name="nopopup"
+  %><jsp:useBean id="currentUser" type="jetbrains.buildServer.users.SUser" scope="request"
+  /><c:set var="linkText" value="None" scope="request"/>
+<c:set var="tags" value="${buildPromotion.tags}"/>
+<c:set var="numberOfTags" value="${fn:length(buildPromotion.tags)}"/>
+<c:choose>
+  <c:when test="${numberOfTags==0}">
+    <c:set var="iconClass" value="icon-tag"/><c:set var="iconWrapperClass" value="noTags"/>
+  </c:when>
+  <c:when test="${numberOfTags==1}">
+    <c:set var="iconClass" value="icon-tag"/><c:set var="iconWrapperClass" value="oneTag"/>
+  </c:when>
+  <c:when test="${numberOfTags>1}">
+    <c:set var="iconClass" value="icon-tags"/><c:set var="iconWrapperClass" value="manyTags"/>
+  </c:when>
+</c:choose>
+<c:choose>
+  <c:when test="${numberOfTags==0}">
+    <c:set var="iconClass" value="icon-tag"/><c:set var="iconWrapperClass" value="noTags"/>
+  </c:when>
+  <c:when test="${numberOfTags==1}">
+    <c:set var="iconClass" value="icon-tag"/><c:set var="iconWrapperClass" value="oneTag"/>
+    <tags:printTag tag="${tags[0]}"/>
+  </c:when>
+  <c:when test="${numberOfTags>1}">
+    <c:set var="iconClass" value="icon-tags"/><c:set var="iconWrapperClass" value="manyTags"/>
+    <c:forEach items="${tags}" var="tag">
+      <tags:printTag tag="${tag}"/>
+    </c:forEach>
+  </c:when>
+</c:choose>
+<tags:_editTagsLink iconWrapperClass="${iconWrapperClass}" iconClass="${iconClass}" buildPromotion="${buildPromotion}"/>
Index: web/WEB-INF/tags/tags/_editTagsLink.tag
===================================================================
--- web/WEB-INF/tags/tags/_editTagsLink.tag	(revision Shelved version)
+++ web/WEB-INF/tags/tags/_editTagsLink.tag	(revision Shelved version)
@@ -0,0 +1,27 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"
+  %><%@ taglib prefix="fn" uri="http://java.sun.com/jsp/jstl/functions"
+  %><%@ taglib prefix="bs" tagdir="/WEB-INF/tags"
+  %><%@ taglib prefix="authz" tagdir="/WEB-INF/tags/authz"
+  %><%@ taglib prefix="tags" tagdir="/WEB-INF/tags/tags" %><%@
+  attribute name="buildPromotion" required="true" type="jetbrains.buildServer.serverSide.BuildPromotion"%><%@
+  attribute name="iconClass" required="true"%><%@
+  attribute name="iconWrapperClass" required="true"%><%@
+  attribute name="className" required="false" type="java.lang.String"%>
+<authz:authorize allPermissions="TAG_BUILD" projectId="${buildPromotion.buildType.projectId}">
+<c:set var="tags"><c:forEach var="tag" items="${buildPromotion.tags}" >${tag} </c:forEach></c:set>
+<c:set var="escapedTags"><bs:escapeForJs text="${tags}" forHTMLAttribute="true"/></c:set>
+<jsp:useBean id="currentUser" type="jetbrains.buildServer.users.SUser" scope="request"/>
+<%--<span class="${iconWrapperClass}" onclick="return BS.Tags.showEditDialog(${buildPromotion.id}, '${escapedTags}', '', ${buildPromotion.numberOfDependencies}, 'availableTags_${buildPromotion.id}');">--%>
+<i class="${iconClass} ${iconWrapperClass}" onclick="return BS.Tags.showEditDialog(${buildPromotion.id}, '${escapedTags}', '', ${buildPromotion.numberOfDependencies}, 'availableTags_${buildPromotion.id}');"/>
+<div id="availableTags_${buildPromotion.id}" style="display: none;">
+  <c:set var="tags" value="${buildPromotion.buildType.tags}"/>
+  <c:if test="${fn:length(tags) gt 0}">
+    <div class="tagsContainer">
+      <c:forEach var="tagName" items="${tags}">
+        <c:set var="escapedTag"><bs:escapeForJs text="${tagName}" forHTMLAttribute="true"/></c:set>
+        <tags:printTag tag="${tagName}" markAsPrivate="${false}" onclick="BS.Tags.appendTag('${escapedTag}'); return false"/>
+      </c:forEach>
+    </div>
+  </c:if>
+</div>
+</authz:authorize>
\ No newline at end of file
